### PR TITLE
C8y mapping of entity registration messages

### PR DESCRIPTION
**Pull request by albinsuresh**
## Proposed changes

Cumulocity mapping for:
- [x] Child device registration
- [x] Service registration
- [x] Nested child device/service registration
- [x] System tests

Deferred for later:
- [ ] Inventory updates for custom fragments
- [ ] Auto-registration for services
- [ ] Avoid resending all entity creation messages to c8y on restart

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
<br/>

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by reubenmiller** [comment]
@albinsuresh I've been trying out the new settings,

* `name` is used instead of the `displayName` property defined in the MQTT spec. I'm ok with using `name` over `displayName` however then we need to update the spec to the implementation
* The Cumulocity IoT external identities for both the child devices and services are missing the common name prefix (which reduces naming conflicts between child devices on different tedge instances).
    * **child-device:** `device:TST_jawbone_direct_wood` -> `<common_name>:device:TST_jawbone_direct_wood`
    * **service:** `device:TST_jawbone_direct_wood:service:custom-app` -> `<common_name>:device:TST_jawbone_direct_wood:service:custom-app`

**Comment by albinsuresh** [comment]
> * `name` is used instead of the `displayName` property defined in the MQTT spec. I'm ok with using `name` over `displayName` however then we need to update the spec to the implementation

Stuck with `name` over `displayName` as thin-edge on its own doesn't really have a "display" aspect to it by default, and that word felt more appropriate for the UI aspects of the cloud. So, `name` sounds more "agnostics" to my ears ;) And it was more consistent with `type` as well. Updated the specs accordingly.

> * The Cumulocity IoT external identities for both the child devices and services are missing the common name prefix (which reduces naming conflicts between child devices on different tedge instances).
>   
>   * **child-device:** `device:TST_jawbone_direct_wood` -> `<common_name>:device:TST_jawbone_direct_wood`
>   * **service:** `device:TST_jawbone_direct_wood:service:custom-app` -> `<common_name>:device:TST_jawbone_direct_wood:service:custom-app`

Fixed by appending the the main device's `device_id` to the generated external IDs of entities.

File: crates/extensions/c8y_mapper_ext/src/converter.rs
Comment (albinsuresh) [inline crates/extensions/c8y_mapper_ext/src/converter.rs:177]: @didier-wenzek and I have agreed to avoid the duplication of these maps here and reuse the mapping in the entity store itself for the same.

The key thing is that, the mapping of `EntityTopicId` to its C8y representation would be done at the converter level, as it is a cloud specific logic, but the mapped id will be used inside the entity store as the `id` of the entity. From the entity store's perspective, it's just one unique opaque ID.  That way, we have clear separation of concerns. So, when different kinds of mappers/components use the `EntityStore` data structure, they are free to do different kinds of `id` mappings and store those mapped ids in the same struct, without needing yet another external mapping.

File: crates/extensions/c8y_mapper_ext/src/converter.rs
Comment (albinsuresh) [inline crates/extensions/c8y_mapper_ext/src/converter.rs:264]: Will create a new helper function over this piece of code that'll return the target entity's C8y topic (e.g: `c8y/s/us/entity-1/entity-2/entity-3`) directly from the `EntityTopicId`.

File: crates/extensions/c8y_mapper_ext/src/converter.rs
Comment (didier-wenzek) [inline crates/extensions/c8y_mapper_ext/src/converter.rs:314]: This can be replaced by an `ancestors.iter().map(...).collect()`.

**Review by didier-wenzek — CHANGES_REQUESTED** [review]
The duplications with the `EntityStore` must be resolved.

**Review by Bravo555 — COMMENTED** [review]
No issues with the code, just requesting some additional comments

File: crates/core/c8y_api/src/smartrest/topic.rs
Comment (albinsuresh) [reply crates/core/c8y_api/src/smartrest/topic.rs:152]: Will do the cleanup of the `c8y_api::topic` module in a follow-up PR, as replacing something like `C8yTopic` would touch a lot of files.

File: crates/extensions/c8y_mapper_ext/src/converter.rs
Comment (didier-wenzek) [inline crates/extensions/c8y_mapper_ext/src/converter.rs:787]: It's fragile to let this logic to the mapper. The only piece of code that is related to c8y is the `map_to_c8y_external_id`  method. See my comment on the `entity_store.update()` method.

File: crates/extensions/c8y_mapper_ext/src/converter.rs
Comment (didier-wenzek) [inline crates/extensions/c8y_mapper_ext/src/converter.rs:826]: This are local registration messages. Don't we also to call `self.try_convert_entity_registration()` to register this entity on c8y?

- Is the expectation to have this mapper processing the local registration messages sent by itself?
- If so telemetry data might be sent to c8y before the child device registration.

File: crates/core/tedge_api/src/entity_store.rs
Comment (albinsuresh) [reply crates/core/tedge_api/src/entity_store.rs:143]: It felt more natural, as even in the physical world when you think about your ancestors, you start with your immediate parents, right? That's why I kept it this way.

File: crates/core/tedge_api/src/entity_store.rs
Comment (albinsuresh) [reply crates/core/tedge_api/src/entity_store.rs:223]: Yeah, I was also initially thinking of accepting a closure to compute that external id when needed. But, since the external id computation relies on the state of the converter (the main device id), I thought it wouldn't work. That's why I decided to pass the computed value this way. And yes, it has the downside of unnecessarily computing the external id when an `@id` value is already present in the message. But, because of that closure limitation, I was forced to do it this way.

File: crates/core/tedge_api/src/entity_store.rs
Comment (albinsuresh) [reply crates/core/tedge_api/src/entity_store.rs:173]: It was working because of the way the `external_id` was computed and passed into the entity store on an `update`. That compute logic in the converter was using that `@id` as-is when one is provided. But, this is going to to be restored once the computation logic is passed in as a closure.

File: crates/core/tedge_api/src/entity_store.rs
Comment (rina23q) [reply crates/core/tedge_api/src/entity_store.rs:143]: I feel it natural, starting with the device in focus, then the last is the main device.

File: crates/core/tedge_api/src/entity_store.rs
Comment (albinsuresh) [reply crates/core/tedge_api/src/entity_store.rs:300]: I've made an attempt to simplify it.

File: tests/RobotFramework/tests/cumulocity/registration/device_registration.robot
Comment (albinsuresh) [inline tests/RobotFramework/tests/cumulocity/registration/device_registration.robot:83]: Even though the registration of a new main device can't be supported atm, I'm assuming that the expected behaviour here was to just add this custom topic id also into the entity store, associated to the main device's external id itself so that data to the main device can be sent via this custom topic as well. That much can be done. The `@id`, `name` and `type` fields will be ignored until we support device fragment updates as well.

File: crates/core/tedge_api/src/entity_store.rs
Comment (didier-wenzek) [reply crates/core/tedge_api/src/entity_store.rs:154]: Okay then - maybe just me :-)

File: crates/core/tedge_api/src/entity_store.rs
Comment (albinsuresh) [inline crates/core/tedge_api/src/entity_store.rs:261]: There was a suggestion from @reubenmiller to apply the same transformation rules(replace `/` with `:`) to the user provided `@id` value as well, rather than just using it as-is, which could mess up the external ID in SmartREST topics. I was initially thinking of just treating it as bad input if it contains "special forbidden chars" like `/, +, #`. But, this proposed scheme might be a bit more flexible, with the caveat of us overrriding even user overridden value. So,  @didier-wenzek, thoughts?

File: crates/core/tedge_api/src/entity_store.rs
Comment (didier-wenzek) [reply crates/core/tedge_api/src/entity_store.rs:300]: Indeed simpler and clearer.

I think that we can go a step further by returning a vector of `RegistrationMessage` instead of `Mqtt::Message` (the latter can be easily built from the former by a mapper). => I will push a commit for that.

**Review by didier-wenzek — COMMENTED** [review]
I'm happy to merge this PR once the conflicts fixed.

File: crates/extensions/c8y_mapper_ext/src/tests.rs
Comment (didier-wenzek) [inline crates/extensions/c8y_mapper_ext/src/tests.rs:1109]: Having to fix this test (`mapper_dynamically_updates_supported_operations_for_child_device`) raises questions it would be good to address before merging.

The issue is that the new c8y external ids must be used only over the c8y topics but also for the c8y operation directories. The current behavior might be confusing: the same device being auto-registered differently after an operation file vs on reception of measurement.

**Review by didier-wenzek — CHANGES_REQUESTED** [review]
One needs to check what are the impacts of using the new external ids on the c8y topics.

File: crates/extensions/c8y_mapper_ext/src/tests.rs
Comment (albinsuresh) [reply crates/extensions/c8y_mapper_ext/src/tests.rs:1109]: Even though this directory mechanism won't be around for too long, as long as we support it, it should preserve its former behaviour IMO: where the external user just creates directories with the local device ID, and the mapper should generate the longer external ID with all `<MAIN_DEVICE_ID>:device` prefix. The user shouldn't be forced to create the directories with those prefixes as that is an internal mapper logic.  The child directory name should be treated exactly like the `device_id` in `te/device/<device_id>//` topic.

File: crates/extensions/c8y_mapper_ext/src/converter.rs
Comment (albinsuresh) [reply crates/extensions/c8y_mapper_ext/src/converter.rs]: The `external_device_registration_message` should have been replaced with `try_convert_entity_registration` message as the `external_device_registration_message` is redundant and incomplete(doesn't cover services). The C8y device/service registration should happen from this scope itself, and we can't wait till the mapper processes the auto-generated registration messages on the next attempt, to avoid the telemetry message going to the cloud before its device's registration message.

Although harmless in its current form, we still need to do something to avoid reprocessing those auto-generated messages. But, as mentioned in the description, I'd like to defer that for a follow-up PR, where can handle this case, and the duplicate processing of all the retained messages(including alarms) on a mapper restart with a single mechanism, if possible. I was thinking of persisting some of the reg information on the file system to tackle this. But, that is to be discussed and I'd like to keep that out of the scope of this PR.

File: crates/core/tedge_api/src/entity_store.rs
Comment (albinsuresh) [reply crates/core/tedge_api/src/entity_store.rs:300]: I'll do that, as it'd allow the reuse of `try_convert_entity_registration` fn for both explicit registration as well as auto-registration.

File: crates/extensions/c8y_mapper_ext/src/converter.rs
Comment (didier-wenzek) [reply crates/extensions/c8y_mapper_ext/src/converter.rs]: > The C8y device/service registration should happen from this scope itself, and we can't wait till the mapper processes the auto-generated registration messages on the next attempt, to avoid the telemetry message going to the cloud before its device's registration message.

Correct. So do you plan to update this PR accordingly?

File: tests/RobotFramework/tests/cumulocity/registration/device_registration.robot
Comment (didier-wenzek) [reply tests/RobotFramework/tests/cumulocity/registration/device_registration.robot:83]: > Even though the registration of a new main device can't be supported atm

I think the issue is to create the EntityStore from a main-device registration message: this message can only be a guess of the topic id used by the main device.

The EntityStore should be created only after a main-device CN. It's registration can then be done either explicitly with a registration message or automatically if some MEA is published on `device/main///#`.

File: crates/extensions/c8y_mapper_ext/src/tests.rs
Comment (didier-wenzek) [reply crates/extensions/c8y_mapper_ext/src/tests.rs:1109]: > the external user just creates directories with the local device ID, and the mapper should generate thes  longer external ID with all <MAIN_DEVICE_ID>:device prefix. The user shouldn't be forced to create the directories with those prefixes as that is an internal mapper logic. The child directory name should be treated exactly like the device_id in te/device/<device_id>// topic.

Fully agree. And this needs to be done along this PR.

I think the `children` cache used by the mapper must also be updated to use external-ids for the keys.

File: crates/core/tedge_api/src/entity_store.rs
Comment (didier-wenzek) [inline crates/core/tedge_api/src/entity_store.rs:333]: I find weird the way this `EntityRegistrationMessage` is built, i.e. by parsing a message payload.

Why not directly building an `EntityRegistrationMessage`? This would avoid to:
- `unwrap()`
- Use an hard-coded `MQTT_ROOT`

File: crates/core/tedge_api/src/entity_store.rs
Comment (didier-wenzek) [inline crates/core/tedge_api/src/entity_store.rs:532]: The `MQTT_ROOT` should not be hard-coded.

File: crates/extensions/c8y_mapper_ext/src/actor.rs
Comment (didier-wenzek) [inline crates/extensions/c8y_mapper_ext/src/actor.rs:131]: This is the drawback of using *short* child device names in path: one has to assume a default topic scheme.

Ideally, I would use the same hierarchy for paths as for MQTT. i.e. with a `device` directory and there a sub directory per child plus a `main` directory. Or an `a/b/c/d` tree if the user prefer its own schema.

Comment (didier-wenzek) [reply crates/extensions/c8y_mapper_ext/src/actor.rs:131]: > we could do that in a follow-up PR

Sure.

File: crates/core/c8y_api/src/smartrest/topic.rs
Comment (albinsuresh) [reply crates/core/c8y_api/src/smartrest/topic.rs]: Got rid of that function.

File: crates/core/tedge_api/src/entity_store.rs
Comment (albinsuresh) [reply crates/core/tedge_api/src/entity_store.rs]: Moved the this function to the converter where the schema is available.

File: tests/RobotFramework/tests/cumulocity/registration/device_registration.robot
Comment (albinsuresh) [reply tests/RobotFramework/tests/cumulocity/registration/device_registration.robot]: Enabled overriding of the default schema using another registration message with custom topic scheme.

File: crates/extensions/c8y_mapper_ext/src/converter.rs
Comment (albinsuresh) [reply crates/extensions/c8y_mapper_ext/src/converter.rs]: Done.

**Comment by albinsuresh** [comment]
Follow-up tasks:
- [ ] Sanitise `external_id` provided via `@id` value in the registration message
- [ ] Review file system based child device registration API. Should the directory name be used as-is as the child device external id? How to support addition of more operations when someone is already using the default topic scheme and hence the default external_id generation logic.
- [ ] Avoid re-processing of auto-registered entity registration messages
- [ ] Avoid re-processing of all the retained entity registration messages on mapper restart. Use file system persistence to identify the delta? Or re-use the message sync mechanism of alarms (too complex)?
- [ ] Simplify `smartrest::topic` module by keeping `publish_topic_from_ancestors` and removing redundant helpers

**Review by didier-wenzek — APPROVED** [review]
Approved. Thank you!

**Pull request merged**
